scripts/pre_commit_linter.py

```diff
@@ -1927,6 +1933,99 @@ def _check_for_copyright_notice(self):
+                        failed = True
+                    else:
+                        line_in_concern = line.split('@')[0].strip()
+                        if line_in_concern.endswith('/'):
+                            line_in_concern = line_in_concern[:-1]
+                        if not glob.glob(line_in_concern.replace('/', '', 1)):
+                            print ('%s --> Pattern on line %s doesn\'t match '
+                                   'any file or directory' % (
+                                       codeowner_file, line_num))
+                            failed = True
+                        if (not line_in_concern.startswith('/') and
+                                not './' +
+                                line_in_concern in CODEOWNER_FILE_PATHS):
+                            print ('%s --> Pattern on line %s is invalid. Use '
+                                   'full path relative to the root directory'
+                                   % (codeowner_file, line_num))
+                            failed = True
+                        if '**' in line_in_concern:
+                            print ('%s --> Pattern on line %s is invalid. '
+                                   '\'**\' wildcard not allowed' % (
+                                        codeowner_file, line_num))
+                            failed = True
+                        path_patterns.append(line_in_concern.replace(
+                            '/', '', 1))
+
+            # Checks that every dir/file is covered under CODEOWNERS.
+            for root, _, filename in os.walk('.'):
+                for file_name in filename:
+                    is_root_file = False
+                    if os.path.join(root, file_name) in CODEOWNER_FILE_PATHS:
+                        is_root_file = True
+                    if any(root.startswith(
+                            x) for x in CODEOWNER_DIR_PATHS) or is_root_file:
+                        match = False
+                        if file_name.endswith(
+                                '.pyc') or file_name == '__init__.py':
+                            match = True
+                            continue
+                        for path_to_match in path_patterns:
+                            if path_to_match.split('/')[-1]:
+                                level = len(path_to_match.split('/'))
+                            else:
+                                level = len(path_to_match.split('/')) - 1
+                            if os.path.join(*((
+                                    os.path.join(root, file_name).replace(
+                                        './', '', 1)).split(
+                                            '/')[:level])) in glob.glob(
+                                                path_to_match):
+                                match = True
+                                break
+                        if not match and self.verbose_mode_enabled:
```

Comment:
Why do we need a warning? The check should fail and not warn, right?

Reply:
Actually, there are many files/dirs without codeowners and @DubeySandeep is currently working on it. So until then, just a warning can be issued.
Also if suppose a new file is created by a contributor, he can't be the codeowner for that file and this will fail the tests.

Reply:
Agreed -- thanks for the explanation.